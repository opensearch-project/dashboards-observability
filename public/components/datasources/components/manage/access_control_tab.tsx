/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiFlexGroup,
  EuiFlexItem,
  EuiSpacer,
  EuiText,
  EuiHorizontalRule,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { EuiPanel } from '@elastic/eui';
import { ConnectionManagementCallout } from './connection_management_callout';
import { coreRefs } from '../../../../framework/core_refs';
import { QueryPermissionsConfiguration } from './query_permissions';
import { DATACONNECTIONS_BASE } from '../../../../../common/constants/shared';
import { SaveOrCancel } from '../save_or_cancel';
import { Role } from '../../../../../common/types/data_connections';

interface AccessControlTabProps {
  dataConnection: string;
  connector: string;
  properties: unknown;
  allowedRoles: string[];
}

export const AccessControlTab = (props: AccessControlTabProps) => {
  const [mode, setMode] = useState<'view' | 'edit'>('view');
  const [roles, setRoles] = useState<Role[]>([]);
  const [selectedQueryPermissionRoles, setSelectedQueryPermissionRoles] = useState<Role[]>(
    props.allowedRoles.map((role) => {
      return { label: role };
    })
  );
  const { http } = coreRefs;

  useEffect(() => {
    http!.get('/api/v1/configuration/roles').then((data) =>
      setRoles(
        Object.keys(data.data).map((key) => {
          return { label: key };
        })
      )
    );
  }, []);

  const AccessControlDetails = () => {
    return (
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiFlexGroup direction="column">
            <EuiFlexItem grow={false}>
              <EuiText className="overview-title">Query access</EuiText>
              <EuiText size="s" className="overview-content">
                {selectedQueryPermissionRoles.length ? `Restricted` : '-'}
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const EditAccessControlDetails = () => {
    return (
      <EuiFlexGroup direction="column">
        <QueryPermissionsConfiguration
          roles={roles}
          selectedRoles={selectedQueryPermissionRoles}
          setSelectedRoles={setSelectedQueryPermissionRoles}
          layout={'vertical'}
        />
      </EuiFlexGroup>
    );
  };

  const saveChanges = () => {
    http!.put(`${DATACONNECTIONS_BASE}`, {
      body: JSON.stringify({
        name: props.dataConnection,
        allowedRoles: selectedQueryPermissionRoles.map((role) => role.label),
        connector: props.connector,
        properties: props.properties,
      }),
    });
    setMode('view');
  };

  const AccessControlHeader = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText size="m">
            <h2 className="panel-title">Access Control</h2>
            Control which OpenSearch users have access to this data source.
          </EuiText>
        </EuiFlexItem>

        <EuiFlexItem grow={false}>
          <EuiButton
            data-test-subj="createButton"
            onClick={() => setMode(mode === 'view' ? 'edit' : 'view')}
            fill={mode === 'view' ? true : false}
          >
            {mode === 'view' ? 'Edit' : 'Cancel'}
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  return (
    <>
      <EuiSpacer />
      <ConnectionManagementCallout />
      <EuiSpacer />
      <EuiPanel>
        <AccessControlHeader />
        <EuiHorizontalRule />
        {mode === 'view' ? <AccessControlDetails /> : <EditAccessControlDetails />}
      </EuiPanel>
      <EuiSpacer />
      {mode === 'edit' && (
        <SaveOrCancel
          onCancel={() => {
            setMode('view');
          }}
          onSave={saveChanges}
        />
      )}
      <EuiSpacer />
    </>
  );
};
