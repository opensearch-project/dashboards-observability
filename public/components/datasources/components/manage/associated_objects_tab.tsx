/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useEffect, useState } from 'react';
import {
  EuiInMemoryTable,
  EuiLink,
  EuiPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiText,
  EuiHorizontalRule,
  EuiButton,
  EuiSpacer,
  EuiEmptyPrompt,
} from '@elastic/eui';
import { AssociatedObject } from 'common/types/data_connections';
import { AccelerationsRecommendationCallout } from './accelerations_recommendation_callout';

interface AssociatedObjectsTabProps {
  associatedObjects: AssociatedObject[];
}

export const AssociatedObjectsTab: React.FC<AssociatedObjectsTabProps> = ({
  associatedObjects,
}) => {
  const [lastUpdated, setLastUpdated] = useState('');

  // TODO: FINISH THE REFRESH LOGIC
  const fetchAssociatedObjects = async () => {
    // Placeholder for data fetching logic
    // After fetching data:
    // setAssociatedObjects(fetchedData);
    const now = new Date();
    setLastUpdated(now.toUTCString()); // Update last updated time
  };

  useEffect(() => {
    fetchAssociatedObjects();
  }, []);

  const AssociatedObjectsHeader = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText size="m">
            <h2 className="panel-title">Associated objects</h2>
            Manage objects associated with this data sources.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <div style={{ textAlign: 'right' }}>
            <EuiText color="subdued" style={{ fontSize: 'small' }}>
              Last updated at:
            </EuiText>
            <EuiText color="subdued" style={{ fontSize: 'small' }}>
              {lastUpdated}
            </EuiText>
          </div>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiButton
            data-test-subj="freshButton"
            iconType="refresh"
            onClick={fetchAssociatedObjects}
          >
            Refresh
          </EuiButton>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const noDataMessage = (
    <EuiEmptyPrompt
      title={<h2>You have no associated objects</h2>}
      body={<p>Add or config tables from your data source or use Query Workbench.</p>}
      actions={
        <EuiButton
          color="primary"
          fill
          onClick={() => window.open('https://example.com', '_blank')}
          iconType="popout"
          iconSide="left"
        >
          Query Workbench
        </EuiButton>
      }
    />
  );

  const columns = [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      'data-test-subj': 'nameCell',
      render: (name: string) => (
        <EuiLink href="https://example.com" target="_blank">
          {name}
        </EuiLink>
      ),
    },
    {
      field: 'database',
      name: 'Database',
      truncateText: true,
      render: (database: string) => (
        <EuiLink href="https://example.com" target="_blank">
          {database}
        </EuiLink>
      ),
    },
    {
      field: 'type',
      name: 'Type',
      sortable: true,
    },
    {
      field: 'createdByIntegration',
      name: 'Created by Integration',
      sortable: true,
    },
    {
      field: 'accelerations',
      name: 'Accelerations',
      sortable: true,
    },
    {
      name: 'Actions',
      actions: [
        {
          name: 'Edit',
          description: 'Edit this object',
          type: 'icon',
          icon: 'discoverApp',
          onClick: (item: AssociatedObject) => console.log('Edit', item),
        },
        {
          name: 'Delete',
          description: 'Delete this object',
          type: 'icon',
          icon: 'bolt',
          onClick: (item: AssociatedObject) => console.log('Delete', item),
        },
      ],
    },
  ];

  const search = {
    box: {
      incremental: true,
      schema: {
        fields: { name: { type: 'string' }, database: { type: 'string' } },
      },
    },
  };

  const pagination = {
    initialPageSize: 10,
    pageSizeOptions: [10, 25, 50],
  };

  const sorting = {
    sort: {
      field: 'name',
      direction: 'asc',
    },
  };

  return (
    <>
      <EuiSpacer />
      <EuiPanel>
        <AssociatedObjectsHeader />
        <EuiHorizontalRule />
        <AccelerationsRecommendationCallout />
        <EuiSpacer />
        {associatedObjects.length > 0 ? (
          <EuiInMemoryTable
            items={associatedObjects}
            columns={columns}
            search={search}
            pagination={pagination}
            sorting={sorting}
            noItemsMessage={associatedObjects.length === 0 ? noDataMessage : undefined}
          />
        ) : (
          noDataMessage
        )}
      </EuiPanel>
      <EuiSpacer />
    </>
  );
};
