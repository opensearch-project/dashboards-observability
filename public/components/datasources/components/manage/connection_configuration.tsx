/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiFieldPassword,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiSelect,
  EuiSpacer,
  EuiText,
  EuiTextArea,
} from '@elastic/eui';
import { EuiSelectOption } from '@elastic/eui/src/components/form/select';
import React, { useState } from 'react';

interface ConnectionConfigurationProps {
  connectionName: string;
  connectionDetails: string;
  onConnectionDetailsChange: (e: any) => void;
  authenticationOptions: EuiSelectOption[];
  setSelectedAuthenticationMethod: (authenticationMethod: EuiSelectOption) => void;
  selectedAuthenticationMethod: string;
}

export const ConnectionConfiguration = (props: ConnectionConfigurationProps) => {
  const {
    connectionName,
    connectionDetails,
    onConnectionDetailsChange,
    authenticationOptions,
    selectedAuthenticationMethod,
    setSelectedAuthenticationMethod,
  } = props;
  const [details, setDetails] = useState(connectionDetails);

  const [password, setPassword] = useState('');

  const NameRow = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Data source name</EuiText>
          <EuiText size="s" className="overview-content">
            This is the name of the data source and how it will be referenced in OpenSearch
            Dashboards.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Data source name">
            <EuiFieldText placeholder={connectionName} readOnly />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const SparkEndpointRow = () => {
    return (
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Spark endpoint URL</EuiText>
          <EuiText size="s" className="overview-content">
            {
              "The URL for your Spark cluster and where your data is. This is what OpenSearch will connect to. The endpoint URL can't be changed. If you'd like to use another endpoint create a new data source."
            }
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Spark endpoint URL">
            <EuiFieldText placeholder={'-'} readOnly />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  return (
    <EuiFlexItem>
      <NameRow />
      <EuiSpacer />
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Description - optional</EuiText>
          <EuiText size="s" className="overview-content">
            Text that can help identify the data source or share additional details
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Description">
            <EuiTextArea
              placeholder="Placeholder text"
              aria-label="test"
              value={details}
              onChange={(e) => {
                setDetails(e.target.value);
              }}
              onBlur={onConnectionDetailsChange}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
      <EuiSpacer />
      <SparkEndpointRow />
      <EuiSpacer />
      <EuiFlexGroup direction="row">
        <EuiFlexItem>
          <EuiText className="overview-title">Authentication details</EuiText>
          <EuiText size="s" className="overview-content">
            This is information used to authenticate and create a data source with Spark.
          </EuiText>
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiFormRow label="Authentication method">
            <EuiSelect
              options={authenticationOptions}
              value={selectedAuthenticationMethod}
              onChange={(e) => setSelectedAuthenticationMethod(e)}
            />
          </EuiFormRow>
          <EuiFormRow label="Username">
            <EuiFieldText placeholder={'Username placeholder'} />
          </EuiFormRow>
          <EuiFormRow label="Password">
            <EuiFieldPassword
              type="dual"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>
    </EuiFlexItem>
  );
};
