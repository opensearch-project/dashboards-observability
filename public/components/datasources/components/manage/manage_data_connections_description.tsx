/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiSpacer,
  EuiText,
  EuiTitle,
  EuiHorizontalRule,
  EuiLink,
  EuiSmallButton,
  EuiFlexGroup,
  EuiFlexItem,
} from '@elastic/eui';
import React from 'react';

interface DataConnectionsDescriptionProps {
  refresh: () => void;
}

export const DataConnectionsDescription = (props: DataConnectionsDescriptionProps) => {
  const { refresh } = props;
  return (
    <div>
      <EuiFlexGroup justifyContent="spaceBetween">
        <EuiFlexItem grow={false}>
          <>
            <EuiTitle size="s">
              <h2>Manage data sources</h2>
            </EuiTitle>

            <EuiSpacer size="s" />
            <EuiText size="s" color="subdued">
              Manage existing data sources or{' '}
              <EuiLink href="#/new">create a new data source</EuiLink>
            </EuiText>
          </>
        </EuiFlexItem>
        <EuiFlexItem grow={false}>
          <EuiSmallButton onClick={refresh} iconType="refresh">
            Refresh
          </EuiSmallButton>
        </EuiFlexItem>
      </EuiFlexGroup>

      <EuiHorizontalRule size="full" />
    </div>
  );
};
