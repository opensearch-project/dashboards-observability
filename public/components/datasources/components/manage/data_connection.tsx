/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiAccordion,
  EuiCard,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiPage,
  EuiPageBody,
  EuiPageHeader,
  EuiPageHeaderSection,
  EuiPanel,
  EuiSpacer,
  EuiTabbedContent,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useEffect, useState } from 'react';
import {
  DATACONNECTIONS_BASE,
  INTEGRATIONS_BASE,
  observabilityMetricsID,
} from '../../../../../common/constants/shared';
import {
  DatasourceDetails,
  PrometheusProperties,
} from '../../../../../common/types/data_connections';
import {
  useLoadAccelerationsToCache,
  useLoadDatabasesToCache,
  useLoadTablesToCache,
} from '../../../../../public/framework/catalog_cache/cache_loader';
import { redirectToExplorerS3 } from './associated_objects/utils/associated_objects_tab_utils';
import { coreRefs } from '../../../../framework/core_refs';
import { getRenderCreateAccelerationFlyout } from '../../../../plugin';
import { NoAccess } from '../no_access';
import { AccelerationTable } from './accelerations/acceleration_table';
import { AccessControlTab } from './access_control_tab';
import { AssociatedObjectsTab } from './associated_objects/associated_objects_tab';
import { InactiveDataConnectionCallout } from './inactive_data_connection';
import {
  InstallIntegrationFlyout,
  InstalledIntegrationsTable,
} from './integrations/installed_integrations_table';

const renderCreateAccelerationFlyout = getRenderCreateAccelerationFlyout();

export const DataConnection = (props: { dataSource: string }) => {
  const { dataSource } = props;
  const [datasourceDetails, setDatasourceDetails] = useState<DatasourceDetails>({
    allowedRoles: [],
    name: '',
    description: '',
    connector: 'PROMETHEUS',
    properties: { 'prometheus.uri': 'placeholder' },
    status: 'ACTIVE',
  });
  const [hasAccess, setHasAccess] = useState(true);
  const { http, chrome, application } = coreRefs;
  const [selectedDatabase, setSelectedDatabase] = useState<string>('');

  const {
    loadStatus: databasesLoadStatus,
    startLoading: startLoadingDatabases,
  } = useLoadDatabasesToCache();
  const { loadStatus: tablesLoadStatus, startLoading: startLoadingTables } = useLoadTablesToCache();
  const {
    loadStatus: accelerationsLoadStatus,
    startLoading: startLoadingAccelerations,
  } = useLoadAccelerationsToCache();

  const cacheLoadingHooks = {
    databasesLoadStatus,
    startLoadingDatabases,
    tablesLoadStatus,
    startLoadingTables,
    accelerationsLoadStatus,
    startLoadingAccelerations,
  };

  const [dataSourceIntegrations, setDataSourceIntegrations] = useState(
    [] as IntegrationInstanceResult[]
  );
  const [refreshIntegrationsFlag, setRefreshIntegrationsFlag] = useState(false);
  const refreshInstances = () => setRefreshIntegrationsFlag((prev) => !prev);

  useEffect(() => {
    const searchDataSourcePattern = new RegExp(
      `flint_${_.escapeRegExp(datasourceDetails.name)}_default_.*`
    );
    const findIntegrations = async () => {
      // TODO: we just get all results and filter, ideally we send a filtering query to the API
      // Should still be probably okay until we get cases of 500+ integration instances
      const result: { data: IntegrationInstancesSearchResult } = await http!.get(
        INTEGRATIONS_BASE + `/store`
      );
      if (result.data?.hits) {
        setDataSourceIntegrations(
          result.data.hits.filter((res) => res.dataSource.match(searchDataSourcePattern))
        );
      } else {
        setDataSourceIntegrations([]);
      }
    };
    findIntegrations();
  }, [http, datasourceDetails.name, refreshIntegrationsFlag]);

  const [showIntegrationsFlyout, setShowIntegrationsFlyout] = useState(false);
  const onclickIntegrationsCard = () => {
    setShowIntegrationsFlyout(true);
  };
  const integrationsFlyout = showIntegrationsFlyout ? (
    <InstallIntegrationFlyout
      closeFlyout={() => setShowIntegrationsFlyout(false)}
      datasourceType={datasourceDetails.connector}
      datasourceName={datasourceDetails.name}
    />
  ) : null;

  const onclickAccelerationsCard = () => {
    renderCreateAccelerationFlyout(dataSource);
  };

  const onclickDiscoverCard = () => {
    redirectToExplorerS3(dataSource);
  };

  const DefaultDatasourceCards = () => {
    return (
      <EuiFlexGroup>
        <EuiFlexItem>
          <EuiCard
            icon={<EuiIcon size="xxl" type="integrationGeneral" />}
            title={'Configure Integrations'}
            description="Connect to common application log types using integrations"
            onClick={onclickIntegrationsCard}
            selectable={{
              onClick: onclickIntegrationsCard,
              isDisabled: false,
              children: 'Add Integrations',
            }}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiCard
            icon={<EuiIcon size="xxl" type="bolt" />}
            title={'Accelerate performance'}
            description="Accelerate query performance through OpenSearch indexing"
            onClick={onclickAccelerationsCard}
            selectable={{
              onClick: onclickAccelerationsCard,
              isDisabled: false,
              children: 'Accelerate Performance',
            }}
          />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiCard
            icon={<EuiIcon size="xxl" type="discoverApp" />}
            title={'Query data'}
            description="Uncover insights from your data or better understand it"
            onClick={onclickDiscoverCard}
            selectable={{
              onClick: onclickDiscoverCard,
              isDisabled: false,
              children: 'Query in Observability Logs',
            }}
          />
        </EuiFlexItem>
      </EuiFlexGroup>
    );
  };

  const fetchSelectedDatasource = () => {
    http!
      .get(`${DATACONNECTIONS_BASE}/${dataSource}`)
      .then((data) => {
        setDatasourceDetails({
          allowedRoles: data.allowedRoles,
          description: data.description,
          name: data.name,
          connector: data.connector,
          properties: data.properties,
          status: data.status,
        });
      })
      .catch((_err) => {
        setHasAccess(false);
      });
  };

  useEffect(() => {
    chrome!.setBreadcrumbs([
      {
        text: 'Data sources',
        href: '#/',
      },
      {
        text: `${dataSource}`,
        href: `#/manage/${dataSource}`,
      },
    ]);
    fetchSelectedDatasource();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [chrome, http]);

  const genericTabs = [
    {
      id: 'access_control',
      name: 'Access control',
      disabled: false,
      content: (
        <AccessControlTab
          allowedRoles={datasourceDetails.allowedRoles}
          dataConnection={dataSource}
          connector={datasourceDetails.connector}
          properties={datasourceDetails.properties}
          key={JSON.stringify(datasourceDetails.allowedRoles)}
        />
      ),
    },
  ];

  const conditionalTabs =
    datasourceDetails.connector === 'S3GLUE'
      ? [
          {
            id: 'associated_objects',
            name: 'Associated Objects',
            disabled: false,
            content: (
              <AssociatedObjectsTab
                datasource={datasourceDetails}
                cacheLoadingHooks={cacheLoadingHooks}
                selectedDatabase={selectedDatabase}
                setSelectedDatabase={setSelectedDatabase}
              />
            ),
          },
          {
            id: 'acceleration_table',
            name: 'Accelerations',
            disabled: false,
            content: (
              <AccelerationTable
                dataSourceName={dataSource}
                cacheLoadingHooks={cacheLoadingHooks}
              />
            ),
          },
          {
            id: 'installed_integrations',
            name: 'Installed Integrations',
            disabled: false,
            content: (
              <InstalledIntegrationsTable
                integrations={dataSourceIntegrations}
                datasourceType={datasourceDetails.connector}
                datasourceName={datasourceDetails.name}
                refreshInstances={refreshInstances}
              />
            ),
          },
        ]
      : [];

  const tabs = [...conditionalTabs, ...genericTabs];

  const QueryOrAccelerateData = () => {
    switch (datasourceDetails.connector) {
      case 'S3GLUE':
        return <DefaultDatasourceCards />;
      case 'PROMETHEUS':
        // Prometheus does not have acceleration or integrations, and should go to metrics analytics
        return (
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiCard
                icon={<EuiIcon size="xxl" type="discoverApp" />}
                title={'Query data'}
                description="Query your data in Metrics Analytics."
                onClick={() => application!.navigateToApp(observabilityMetricsID)}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
        );
      default:
        return <DefaultDatasourceCards />;
    }
  };

  const S3DatasourceOverview = () => {
    return (
      <EuiPanel>
        <EuiFlexGroup>
          <EuiFlexItem>
            <EuiFlexGroup direction="column">
              <EuiFlexItem grow={false}>
                <EuiText className="overview-title">Description</EuiText>
                <EuiText size="s" className="overview-content">
                  {datasourceDetails.description || '-'}
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiFlexGroup direction="column">
              <EuiFlexItem grow={false}>
                <EuiText className="overview-title">Query Access</EuiText>
                <EuiText size="s" className="overview-content">
                  {datasourceDetails.allowedRoles.length > 0
                    ? `Restricted to ${datasourceDetails.allowedRoles.join(', ')}`
                    : 'Admin only'}
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer />
      </EuiPanel>
    );
  };

  const PrometheusDatasourceOverview = () => {
    return (
      <EuiPanel>
        <EuiFlexGroup>
          <EuiFlexItem>
            <EuiFlexGroup direction="column">
              <EuiFlexItem grow={false}>
                <EuiText className="overview-title">Connection title</EuiText>
                <EuiText size="s" className="overview-content">
                  {datasourceDetails.name || '-'}
                </EuiText>
              </EuiFlexItem>
              <EuiFlexItem grow={false}>
                <EuiText className="overview-title">Data source description</EuiText>
                <EuiText size="s" className="overview-content">
                  {datasourceDetails.description || '-'}
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiFlexItem>
          <EuiFlexItem>
            <EuiFlexGroup direction="column">
              <EuiFlexItem grow={false}>
                <EuiText className="overview-title">Prometheus URI</EuiText>
                <EuiText size="s" className="overview-content">
                  {(datasourceDetails.properties as PrometheusProperties)['prometheus.uri'] || '-'}
                </EuiText>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiFlexItem>
        </EuiFlexGroup>
        <EuiSpacer />
      </EuiPanel>
    );
  };

  const DatasourceOverview = () => {
    switch (datasourceDetails.connector) {
      case 'S3GLUE':
        return <S3DatasourceOverview />;
      case 'PROMETHEUS':
        return <PrometheusDatasourceOverview />;
    }
  };

  if (!hasAccess) {
    return <NoAccess />;
  }

  return (
    <EuiPage>
      <EuiPageBody>
        <EuiPageHeader style={{ justifyContent: 'spaceBetween' }}>
          <EuiPageHeaderSection style={{ width: '100%', justifyContent: 'space-between' }}>
            <EuiFlexGroup>
              <EuiFlexItem grow={false}>
                <EuiTitle data-test-subj="dataspirceTitle" size="l">
                  <h1>{dataSource}</h1>
                </EuiTitle>
              </EuiFlexItem>
            </EuiFlexGroup>
          </EuiPageHeaderSection>
        </EuiPageHeader>

        <DatasourceOverview />
        <EuiSpacer />

        {integrationsFlyout}

        {datasourceDetails.status !== 'ACTIVE' ? (
          <InactiveDataConnectionCallout
            datasourceDetails={datasourceDetails}
            fetchSelectedDatasource={fetchSelectedDatasource}
          />
        ) : (
          <>
            <EuiAccordion
              id="queryOrAccelerateAccordion"
              buttonContent={'Get started'}
              initialIsOpen={true}
              paddingSize="m"
            >
              <QueryOrAccelerateData />
            </EuiAccordion>
            <EuiTabbedContent tabs={tabs} />
          </>
        )}

        <EuiSpacer />
      </EuiPageBody>
    </EuiPage>
  );
};
export { DatasourceDetails };
