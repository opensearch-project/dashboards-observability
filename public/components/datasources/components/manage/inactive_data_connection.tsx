/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButton, EuiCallOut } from '@elastic/eui';
import React from 'react';
import {
  DATACONNECTIONS_BASE,
  DATACONNECTIONS_UPDATE_STATUS,
  EDIT,
} from '../../../../../common/constants/shared';
import { coreRefs } from '../../../../framework/core_refs';
import { useToast } from '../../../common/toast';
import { DatasourceDetails } from './data_connection';

interface InactiveDataConnectionCalloutProps {
  datasourceDetails: DatasourceDetails;
  fetchSelectedDatasource: () => void;
}

export const InactiveDataConnectionCallout = ({
  datasourceDetails,
  fetchSelectedDatasource,
}: InactiveDataConnectionCalloutProps) => {
  const { setToast } = useToast();
  const { http } = coreRefs;
  const enableDataSource = () => {
    http!
      .post(`${DATACONNECTIONS_BASE}${EDIT}${DATACONNECTIONS_UPDATE_STATUS}`, {
        body: JSON.stringify({ name: datasourceDetails.name, status: 'active' }),
      })
      .then(() => {
        setToast(`Data connection ${datasourceDetails.name} enabled successfully`);
        fetchSelectedDatasource();
      })
      .catch((err) => {
        console.error(err);
        setToast(`Data connection ${datasourceDetails.name} could not be enabled.`, 'danger');
      });
  };

  return (
    <EuiCallOut title="This data source connection is inactive" color="warning" iconType="help">
      <p>
        Associated objects and accelerations are not available while this connection is inactive.
      </p>
      <EuiButton onClick={enableDataSource} color="warning">
        Enable connection
      </EuiButton>
    </EuiCallOut>
  );
};
