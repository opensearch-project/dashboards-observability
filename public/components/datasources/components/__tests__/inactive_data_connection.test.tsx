/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { act } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import ReactDOM from 'react-dom';
import { coreRefs } from '../../../../../public/framework/core_refs';
import { describePrometheusDataConnection } from '../../../../../test/datasources';
import { DatasourceDetails } from '../manage/data_connection';
import { InactiveDataConnectionCallout } from '../manage/inactive_data_connection';

jest.mock('../../../../../public/framework/core_refs', () => ({
  coreRefs: {
    chrome: {
      setBreadcrumbs: jest.fn(),
    },
    http: {
      get: jest.fn(),
    },
  },
}));

describe('Data Connection Inactive Page test', () => {
  configure({ adapter: new Adapter() });

  beforeEach(() => {
    // Clear the mock implementation before each test
    (coreRefs.http!.get as jest.Mock).mockClear();
  });

  it('Renders inactive data connection callout', async () => {
    const mockDatasourceDetails: DatasourceDetails = {
      allowedRoles: [],
      name: '',
      description: '',
      connector: 'PROMETHEUS',
      properties: { 'prometheus.uri': 'placeholder' },
      status: 'DISABLED',
    };

    const container = document.createElement('div');
    (coreRefs.http!.get as jest.Mock).mockResolvedValue(describePrometheusDataConnection);
    await act(() => {
      ReactDOM.render(
        <InactiveDataConnectionCallout
          datasourceDetails={mockDatasourceDetails}
          fetchSelectedDatasource={jest.fn()}
        />,
        container
      );
    });
    expect(container).toMatchSnapshot();
  });
});
