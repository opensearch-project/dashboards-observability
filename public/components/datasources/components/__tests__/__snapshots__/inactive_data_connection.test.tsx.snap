// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data Connection Inactive Page test Renders inactive data connection callout 1`] = `
<div>
  <div
    class="euiCallOut euiCallOut--warning"
  >
    <div
      class="euiCallOutHeader"
    >
      <svg
        aria-hidden="true"
        class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiCallOutHeader__icon"
        focusable="false"
        height="16"
        role="img"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
        />
      </svg>
      <span
        class="euiCallOutHeader__title"
      >
        This data source connection is inactive
      </span>
    </div>
    <div
      class="euiText euiText--small"
    >
      <div
        class="euiTextColor euiTextColor--default"
      >
        <p>
          Associated objects and accelerations are not available while this connection is inactive.
        </p>
        <button
          class="euiButton euiButton--warning"
          type="button"
        >
          <span
            class="euiButtonContent euiButton__content"
          >
            <span
              class="euiButton__text"
            >
              Enable connection
            </span>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
